apps/emqx_postgresql/src/emqx_postgresql.erl

```diff
@@ -148,13 +153,137 @@ on_start(
             {error, Reason}
     end.
 
-on_stop(InstId, _State) ->
+on_stop(InstId, State) ->
     ?SLOG(info, #{
         msg => "stopping_postgresql_connector",
         connector => InstId
     }),
+    close_connections(State),
     emqx_resource_pool:stop(InstId).
 
+close_connections(#{pool_name := PoolName} = _State) ->
+    WorkerPids = [Worker || {_WorkerName, Worker} <- ecpool:workers(PoolName)],
+    close_connections_with_worker_pids(WorkerPids),
+    ok.
+
+close_connections_with_worker_pids([WorkerPid | Rest]) ->
+    %% We ignore errors since any error probably means that the
+    %% connection is closed already.
+    try ecpool_worker:client(WorkerPid) of
+        {ok, Conn} ->
+            _ = epgsql:close(Conn),
+            close_connections_with_worker_pids(Rest);
+        _ ->
+            close_connections_with_worker_pids(Rest)
+    catch
+        _:_ ->
+            close_connections_with_worker_pids(Rest)
+    end;
+close_connections_with_worker_pids([]) ->
+    ok.
+
+on_add_channel(
+    _InstId,
+    #{
+        installed_channels := InstalledChannels
+    } = OldState,
+    ChannelId,
+    ChannelConfig
+) ->
+    %% The following will throw an exception if the bridge producers fails to start
+    {ok, ChannelState} = create_channel_state(ChannelId, OldState, ChannelConfig),
+    case ChannelState of
+        #{prepares := {error, Reason}} ->
+            {error, {unhealthy_target, Reason}};
+        _ ->
+            NewInstalledChannels = maps:put(ChannelId, ChannelState, InstalledChannels),
+            %% Update state
+            NewState = OldState#{installed_channels => NewInstalledChannels},
+            {ok, NewState}
+    end.
+
+create_channel_state(
+    ChannelId,
+    #{pool_name := PoolName} = _ConnectorState,
+    #{parameters := Parameters} = _ChannelConfig
+) ->
+    State1 = parse_prepare_sql(Parameters, ChannelId),
+    {ok,
+        init_prepare(State1#{
+            pool_name => PoolName,
+            prepare_statement => #{}
+        })}.
+
+on_remove_channel(
+    _InstId,
+    #{
+        installed_channels := InstalledChannels
+    } = OldState,
+    ChannelId
+) ->
+    %% Close prepared statements
+    ok = close_prepared_statement(ChannelId, OldState),
+    NewInstalledChannels = maps:remove(ChannelId, InstalledChannels),
+    %% Update state
+    NewState = OldState#{installed_channels => NewInstalledChannels},
+    {ok, NewState}.
+
+close_prepared_statement(ChannelId, #{pool_name := PoolName} = State) ->
+    WorkerPids = [Worker || {_WorkerName, Worker} <- ecpool:workers(PoolName)],
+    close_prepared_statement(WorkerPids, ChannelId, State),
```

Comment:
Nit: if `WorkerPids = []`, it won't do anything.

Reply:
If there are no worker pids then there are no prepared statements to close.